schema/default.go

```diff
@@ -0,0 +1,150 @@
+}
+
+func getDefaultValue(fieldType FieldType) interface{} {
+	switch fieldType {
+	case BoolType:
+		return false
+	case Int32Type, Int64Type, DoubleType:
+		return 0
+	}
+
+	return nil
+}
+
+type FieldDefaulter struct {
+	value     interface{}
+	createdAt bool
+	updatedAt bool
+}
+
+func newDefaulter(createdAt *bool, updatedAt *bool, name string, dataType FieldType, v interface{}) (*FieldDefaulter, error) {
+	if createdAt != nil && *createdAt {
+		return &FieldDefaulter{
+			createdAt: true,
+		}, nil
+	}
+	if updatedAt != nil && *updatedAt {
+		return &FieldDefaulter{
+			updatedAt: true,
+		}, nil
+	}
+
+	defaulter := &FieldDefaulter{}
+	if v == "" {
+		defaulter.value = getDefaultValue(dataType)
+		return defaulter, nil
+	}
+
+	var err error
+	switch ty := v.(type) {
+	case bool:
+		if dataType != BoolType {
+			return nil, errors.InvalidArgument("default value should be true/false only for boolean field '%s'", name)
+		}
+	case json.Number:
+		if dataType == Int32Type || dataType == Int64Type {
+			if _, err = ty.Int64(); err != nil {
+				return nil, errors.InvalidArgument("default value should be numeric for numeric field '%s'", name)
+			}
+		}
+		if dataType == DoubleType {
+			if _, err = ty.Float64(); err != nil {
+				return nil, errors.InvalidArgument("default value should be numeric for numeric field '%s' %s", name, err.Error())
+			}
+		}
+	case string:
+		if dataType != StringType && dataType != DateTimeType && dataType != UUIDType && dataType != ByteType {
+			return nil, errors.InvalidArgument("default value is not supported for '%s' type: '%s'", name, FieldNames[dataType])
+		}
+		if defaulter.createdAt && dataType != DateTimeType {
+			return nil, errors.InvalidArgument("createdAt is only supported for date-time type")
+		}
+		if defaulter.updatedAt && dataType != DateTimeType {
+			return nil, errors.InvalidArgument("updated is only supported for date-time type")
+		}
+		if v.(string) == funcNowName && dataType != DateTimeType {
+			return nil, errors.InvalidArgument("now() is only supported for date-time type")
+		}
+		if !isSupportedDefaultFunction(v.(string)) {
+			return nil, errors.InvalidArgument("'%s' function is not supported", v)
+		}
+	case []any:
+		if dataType != ArrayType {
+			return nil, errors.InvalidArgument("default value is not supported for '%s' type: '%s'", name, FieldNames[dataType])
+		}
+	default:
+		return nil, errors.InvalidArgument("unsupported default value for field '%s' type: '%s'", name, FieldNames[dataType])
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	defaulter.value = v
+
+	return defaulter, nil
+}
+
+func (defaulter *FieldDefaulter) GetValue(createdAt string, updatedAt string) interface{} {
+	if defaulter.createdAt {
+		if len(createdAt) == 0 {
+			return nil
+		}
+		return createdAt
+	}
+	if defaulter.updatedAt {
+		if len(updatedAt) == 0 {
+			return nil
+		}
+
+		return updatedAt
+	}
+
+	switch defaulter.value {
+	case funcNowName:
+		return createdAt
```

Comment:
There is a case on update where the `createdAt` timestamp is set to `""` which means that a `now()` would also be an empty string. I'm guessing this shouldn't happen as the `now()` should only be done on create but is there a case where we could set the incorrect time?

Reply:
So a field can only be tagged with either one of `createdAt`/`default(now()`/`updatedAt`. Now/CreatedAt will be set during insert/replacing a document and the caller guarantees that createdAt is not empty during insert and updatedAt is not empty during an update.

Reply:
The caller should be able to set it to some value in which case that would take precedence, right?

Reply:
How about this, we only have `default(now()`? Then if you don't specify a value for the fields that are tagged they will be automatically set to the value of `now()`. However, if the user specifies a value when inserting or updating then nothing happens.

Reply:
Postgres and MySQL allow doing something like this:

```
CREATE TABLE note(
    note_id serial PRIMARY KEY,
    message varchar(255) NOT NULL,
    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
    updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);
```

In this case, the value for the fields is populated during insertion time.
In the case of MySQL, you can additionally say:
```
updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP
```

In this case, the value is updated whenever the row is updated.

Reply:
But if the user supplies a value for these columns during insert or update time, that takes precedence.

Reply:
So we are not setting value if the payload has a value already set by the user. That applies to all the functions/defaults/createdAt/updatedAt supported by us. Only set if it is empty in the payload.